### PR TITLE
Run AR validate and hmis_validate earlier in form submission mutations

### PR DESCRIPTION
**Pull request by gigxz**
#### In this PR:
* Run processor before returning errors, so that we don't have to duplicate ALL validations in the FormDefinition.
* Fix issues with resolving AR errors
* Add IncomeBenefit Validator for validating HUD requirements, with new errors around Income/Benefits/Insurance
* Add Enrollment Validator for validating HUD requirements

#### Form validation approach

I think the slightly modified approach (explained below) is an improvement for (1) maintenance cost and (2) ensuring HUD compliance.

I've attempted to reduce the need for duplicate validation logic in both code AND FormDefinitions. Defining HUD-compliance-constraints in FormDefinitions should be avoided –– and we definitely should NOT rely on FormDefinition validation _exclusively_ for any HUD-compliance validations. FormDefinition validation is more error-prone, since  it depends on a well-formed definition, accurate rendering, accurate data transformation by the frontend, and accurate validation by the backend.

* We should use **ActiveRecord Validation** for validations that are absolutely required by HUD, and should always be validated
  * Example: _“If there is a "Yes" response to 'Covered by Health Insurance' then at least one source of health insurance must be identified.”_ (HUD requirement)
  * If a visual treatment is desired, the constraint should be duplicated in the FormDefinition (e.g. specifying that an item is 'required').
* We should use `hmis_validate` in AR Validators (I'm calling this _Custom Record Validation_) for validations and warnings that we want to apply across all installations. We can define them in code (rather than FormDefinitions) if we are sure that they shouldn't change. It also makes sense to use this for warnings/errors that need to fetch a bunch of related data (like counting open enrollments) so that we don't have to thread as much data through the FormDefinition.
  * Example: _Warn if closing a project with open enrollments_; _Warn if Entry Date is equal to DOB_
* Finally, we should use the **FormDefinition** for validations that may need to be configurable per installation. We should also use it for validations that need a visual cue (requiredness) or component impact (date bounds) -- in these cases, it may still be duplicated in the validators.

### CODE REVIEW

File: drivers/hmis/spec/models/hmis/form/form_processor_spec.rb [hunk 1]
```diff
@@ -36,219 +36,292 @@
     expect(assessment.enrollment.enrollment_cocs.first.coc_code).to eq('MA-507')
   end
 
-  it 'ingests IncomeBenefit into the hud tables (income sources)' do
```
Comment (gigxz) [inline drivers/hmis/spec/models/hmis/form/form_processor_spec.rb:39]: Most of the diff in this file is just changing nesting structure, and adding additional tests.

File: drivers/hmis/app/models/hmis/hud/validators/exit_validator.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  def configuration
+    Hmis::Hud::Exit.hmis_configuration(version: '2022').except(*IGNORED)
+  end
+
+  def self.hmis_validate(record, role: nil, **_)
+    errors = HmisErrors::Errors.new
+
+    if record.exit_date.present? && role == :EXIT
+      errors.add :exit_date, :out_of_range, message: 'cannot be in the future' if record.exit_date.future?
+      errors.add :exit_date, :information, severity: :warning, message: 'is over 30 days ago' if record.exit_date < (Date.today - 30.days)
```
Comment (eanders) [inline drivers/hmis/app/models/hmis/hud/validators/exit_validator.rb]: We should add one that it can't be before the entry date

Comment (eanders) [reply drivers/hmis/app/models/hmis/hud/validators/exit_validator.rb]: Oh, that's below.  Never mind

**Review by eanders — APPROVED** [review]
This looks like an improvement to me!

**Comment by gigxz** [comment]
Cool @eanders thank you! Sorry, I pushed another commit before I realized you had reviewed! https://github.com/greenriver/hmis-warehouse/pull/2888/commits/feb01cbe865814e852e2ae603a631a737a4b0118 does a couple things:
1. Call the validator functions from `find_and_validate_assessment_date`
2. Update the `CreateEnrollment` mutation to validate the entry date using the same validator funciton. I got off-topic here and also removed the "in progress" input, since we don't use it. This needs some more work (TODOs/FIXMEs) but I can leave that to another PR.


Edit: the `AddHouseholdMembersToEnrollment` mutation also needs to be updated to do the correct entry date validation. Again, I can do that in the follow-up PR

File: drivers/hmis/app/models/hmis/hud/validators/enrollment_validator.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    errors.add :entry_date, :out_of_range, message: future_message, **options if entry_date.future?
+    errors.add :entry_date, :out_of_range, message: over_twenty_years_ago_message, **options if entry_date < (Date.today - 20.years)
+    errors.add :entry_date, :out_of_range, message: before_dob_message, **options if dob.present? && dob > entry_date
+    errors.add :entry_date, :out_of_range, message: after_exit_message(exit_date), **options if exit_date.present? && exit_date < entry_date
+
+    unless enrollment.head_of_household?
+      hoh_entry_date = enrollment.hoh_entry_date
+      errors.add :entry_date, :out_of_range, severity: :warning, message: before_hoh_entry_message(hoh_entry_date), **options if hoh_entry_date.present? && entry_date < hoh_entry_date
+    end
```
Comment (gigxz) [inline drivers/hmis/app/models/hmis/hud/validators/enrollment_validator.rb:28]: This warns "entry date cannot be before HoH entry date."

It should probably be an error IF we are validating in the context of (1) a new household member being added (ie a new WIP enrollment), or (2) a WIP intake assessment being saved or submitted for the first time.

Comment (eanders) [reply drivers/hmis/app/models/hmis/hud/validators/enrollment_validator.rb:28]: Maybe, though isn't it a problem whenever it happens?

File: drivers/hmis/app/models/hmis/hud/validators/exit_validator.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+  def self.hmis_validate(record, role: nil, **_)
+    errors = HmisErrors::Errors.new
+    errors.push(*validate_exit_date(record, record.exit_date)) if role == :EXIT
+    errors.errors
+  end
+
+  def validate(record)
+    super(record) do
+      record.errors.add :other_destination, :required if record.destination == 17 && !record.other_destination.present?
```
Comment (eanders) [inline drivers/hmis/app/models/hmis/hud/validators/exit_validator.rb]: 17 should probably be pulled out into a constant or method

File: drivers/hmis/app/models/hmis/hud/validators/enrollment_validator.rb (hunk 3 above)
Comment (gigxz) [reply drivers/hmis/app/models/hmis/hud/validators/enrollment_validator.rb:28]: Is it? If the HoH changed, then the old HoH's entry date would be before the HoH's entry date. If you are going back and doing data correction in that persons intake assessment, you shouldn't get an error (I think).
This is still assuming that we are _not_ doing entry date back-dating when the HoH changes.

Comment (eanders) [reply drivers/hmis/app/models/hmis/hud/validators/enrollment_validator.rb:28]: I'm probably not following completely, but current guidance would be that the HoH Entry Date must be on or before any other members.  
As you mentioned, if we assume we aren't back-dating, then it should _not_ be an error, but the guidance is still that the HoH must be there for the full enrollment time.

Comment (gigxz) [reply drivers/hmis/app/models/hmis/hud/validators/enrollment_validator.rb:28]: Right. That's why I made it be a _warning_, not an error. I was just pointing out that in certain cases, we probably do want to treat it as an error. It is something I am planning to come back to for 184407413, I hadn't nailed down the logic well enough to implement it here

Comment (eanders) [reply drivers/hmis/app/models/hmis/hud/validators/enrollment_validator.rb:28]: Sounds good, and sorry, I was using error and warning interchangeably.

**Pull request merged**